# MIMIR
**MIMIR** is an approach for **M**odule **I**dentification via **M**ulti-source **I**ntegration for **R**econstructing Differentiation. Briefly, this approach groups genes into modules using both gene expression similarities from single-cell RNA sequencing (scRNA-seq) data and functional similarities from functional annotation databases. Each gene module contains genes that likely work together to drive a biological process during cell differentiation. MIMIR is part of our effort to extract biological insights from single-cell transcriptomic atlases and trajectories to understand how differentiation is regulated.
## Inputs
1. scRNA-seq data from different stages of cell differentiation. In our example, we use the zebrafish notochord and hatching gland single-cell transcriptomes extracted from datasets published in [Farrell et al. 2018](https://pubmed.ncbi.nlm.nih.gov/29700225/) and [Raj et al. 2020](https://pubmed.ncbi.nlm.nih.gov/33068532/), which covers 17 time points from 3 to 24 hours post fertilization (hpf), during which the notochord and hatching gland cells are specified and differentiate into functionally and structurally specialized cells. 
2. Functional annotations of genes highly expressed during differentiation. In our example, we use annotations from [Gene Ontology](https://geneontology.org/), and [Reactome](https://reactome.org/), [InterPro](https://www.ebi.ac.uk/interpro/). We additionally use the [STRING database](https://string-db.org/cgi/download?sessionId=bykC2Can3gR6), which provides protein-protein interaction predictions.
## Steps in Module Identification
MIMIR involves 5 major steps:
1. Identify enriched genes during differentiation
2. Calculate expression similarities between enriched genes
3. Calculate functional similarities between enriched genes
4. Cluster enriched genes using combined expression and functional similarities
5. Annotation and curate the resulting gene modules
### Step 1: Identify enriched genes during differentiation
Genes that are more highly expressed in the cell type of interest compared to background cell types can be identified using appropriate differential expression tests. In our [example](https://github.com/YiqunW/MIMIR/blob/main/example_scripts/step1_Identify_Enriched_Genes.md), we segmented the notochord transcriptional trajectory into small time windows, and identified genes that show stronger expression in the notochord compared to the rest of the embryo in each time window using a precision-recall curve (AUCPR).
### Step 2: Calculate expression similarities between enriched genes
Expression similarities between enriched genes are calculated by comparing their sc-expression patterns during differentiation. We tested a range of expression distance metrics in our [example](https://github.com/YiqunW/MIMIR/blob/main/example_scripts/step2_Calculate_expression_similarities.md), including Euclidean distance, (soft) cosine distance, Canberra distance, and Jensen-Shannon divergence. 
### Step 3: Calculate functional similarities between enriched genes
Functional annotation databases such as [Gene Ontology](https://geneontology.org/), [Reactome](https://reactome.org/), [InterPro](https://www.ebi.ac.uk/interpro/), and [HGNC](https://www.genenames.org/) each contains annotation terms that are hierarchically organized. We derived genes' functional similarities from the semantic similarities between their annotations. The idea behind semantic similarity between annotation terms is that annotations for similar functions should have a close relationship in the hierarchical structure. Algorithms in the R package [GOSemSim](https://bioconductor.org/packages/release/bioc/html/GOSemSim.html) were adapted to calculate the semantic similarities between funcitonal annotations. The [STRING database](https://string-db.org/cgi/download?sessionId=bykC2Can3gR6) provides protein-protein interaction likelihood scores between genes, which were used directly as similarity scores. Functional similarities calculated from different databases were combined following [the framework STRING database uses](https://string-db.org/cgi/help?sessionId=bYM4qN6d8EXf) for combining scores from different evidence channels. See our [example](https://github.com/YiqunW/MIMIR/blob/main/example_scripts/step3_Calculate_functional_similarities.md) for how we calculate funcitonal similarities between notochord enriched genes.
### Step 4: Cluster enriched genes using combined expression and functional similarities
Expression and functional similarities can be combined using various methods. In our [example](https://github.com/YiqunW/MIMIR/blob/main/example_scripts/step4_Cluster_genes_with_combined_similarities.md), a simple multiplication between scaled expression and functional similarities worked well. Leiden clustering can be used to cluster enriched genes into initial modules.
### Step 5: Annotate and curate the resulting gene modules
We choose a clustering result to serve as the basis of the gene modules (see [example](https://github.com/YiqunW/MIMIR/blob/main/example_scripts/step5_Check_clustering_results.md)). We then use a self-built [Shiny app]() to facilitate the inspection, annotation, and curation of the clustering results to produce the final gene modules. 
